README.rst
===============================
passwordmetrics
===============================

Features
--------

The passwordmetrics module takes a string and gives you a lot of information
about it pertaining it's usefulness as a password.

* Free software: MIT license
* Documentation: http://passwordmetrics.readthedocs.org.

Features
--------

* Gives you a lot of information about a password, such as:
  * Words used in the password
  * Entropy (ie, how hard it is to guess)
* It's configurable for different languages.
* A big english wordlist included.
* L33t compatibility


HISTORY.rst
.. :changelog:

=======
History
=======

1.1 (unreleased)
----------------

- Nothing changed yet.


1.0 (2017-04-15)
----------------

- Stable release, better docs.


1.0b1 (2015-04-29)
------------------

* First release on PyPI.


tests/test_docs.py
import sys
import re
import doctest
import manuel.doctest
import manuel.codeblock
import manuel.testing
import unittest


if sys.version_info[0] < 3:
    # Just don't do them under Python 3.
    # Sigh.

    class CustomChecker(doctest.OutputChecker):

        def check_output(self, want, got, optionflags):
            got = re.sub("set\(\[([^\]]+?)\]\)", "{\\1}", got)
            got = re.sub("set\(\[]\)", "set()", got)
            return doctest.OutputChecker.check_output(self, want, got, optionflags)

        def additional_tests():
            m = manuel.doctest.Manuel(optionflags=doctest.NORMALIZE_WHITESPACE,
                                      checker=CustomChecker())
            m += manuel.codeblock.Manuel()
            return manuel.testing.TestSuite(m, r'../docs/usage.rst')

if __name__ == '__main__':
    unittest.TextTestRunner().run(additional_tests())


tests/test_passwordmetrics.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_passwordmetrics
----------------------------------

Tests for `passwordmetrics` module.
"""
from __future__ import unicode_literals
import unittest
import passwordmetrics
import string
from io import open

class TestPasswordMetrics(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        passwordmetrics.configure()
        cls.words = {'correct': 1, 'horse': 2, 'battery': 3, 'staple': 4, 'a': 1, 'incorrect': 1, 'bat': 1}

    def test_wordcount(self):
        words, rest = passwordmetrics._find_words('correcthorseb!wdbatterystaplerWd3t', self.words)
        self.assertEqual(words, set(['battery', 'horse', 'correct', 'staple']))
        self.assertEqual(rest, 'b!wdrWd3t')

        words, rest = passwordmetrics._find_words('abcdefgh', self.words)
        self.assertEqual(words, {'a'})
        self.assertEqual(rest, 'bcdefgh')

        # Duplicate words count only as one word:
        words, rest = passwordmetrics._find_words('batterybattery', self.words)
        self.assertEqual(words, {'battery'})


    def test_groups(self):
        groups = passwordmetrics._find_groups('abcdefgh')
        self.assertEqual(groups[0], {'lowercase'})

        groups = passwordmetrics._find_groups('12345678')
        self.assertEqual(groups[0], {'digits'})

        groups = passwordmetrics._find_groups('aAaAaAaA')
        self.assertEqual(groups[0], {'lowercase', 'uppercase'})

        groups = passwordmetrics._find_groups('xyFg98%!')
        self.assertEqual(groups[0], {'lowercase', 'uppercase', 'digits', 'punctuation'})

    def test_character_entropy(self):
        entropy, unkown = passwordmetrics._character_entropy('abcdefgh')
        self.assertAlmostEqual(entropy, 37.6035177451287)

        entropy, unkown = passwordmetrics._character_entropy('12345678')
        self.assertAlmostEqual(entropy, 26.5754247590989)

        # Repeated characters count as one:
        entropy, unkown = passwordmetrics._character_entropy('aAaAaAaA')
        self.assertAlmostEqual(entropy, 11.4008794362821)

        # As good as it gets for an 8 char password
        entropy, unkown = passwordmetrics._character_entropy('xyFg98%!')
        self.assertAlmostEqual(entropy, 52.4367108134211)

    def test_substitutions(self):
        # This long password gets a reasonable score on characters alone
        self.assertAlmostEqual(passwordmetrics._character_entropy('Tr0ub4dor&3')[0], 65.54588851677637)
        # But much worse when considering it uses a word, even though it's misspelled and only
        # appears once in the whole corpus (because I put it there).
        self.assertAlmostEqual(passwordmetrics.metrics('Tr0ub4dor&3')['entropy'], 42.11714046349914)
        # Although still better than if there was no substitutions in the word
        self.assertAlmostEqual(passwordmetrics.metrics('Troubador&3')['entropy'], 31.332505617941614)

        words, rest = passwordmetrics._find_words('batterybattery', self.words)
        self.assertEquals(rest, '')

        words, rest = passwordmetrics._find_words('b4tteryb4ttery', self.words)
        self.assertEquals(rest, '4')

        words, rest = passwordmetrics._find_words('b4ttery8attery', self.words)
        self.assertEquals(rest, '48')

    def test_password_entropy(self):
        self.assertAlmostEqual(passwordmetrics.metrics('abcdefgh')['entropy'], 36.7646671442273)
        self.assertAlmostEqual(passwordmetrics.metrics('xyFg98%!')['entropy'], 55.49709479680921)
        self.assertAlmostEqual(passwordmetrics.metrics('correcthorsebatterystaple')['entropy'], 45.58216824444706)

    def test_password_metrics(self):
        metrics = passwordmetrics.metrics('correcthorseb!wdbatterystaplerWd3t')
        self.assertEqual(metrics['words'], {'et', 'stapler', 'i', 'correct', 'battery', 'horse'})
        self.assertEqual(metrics['used_groups'], {'uppercase', 'lowercase', 'punctuation', 'digits'})
        self.assertEqual(metrics['unused_groups'], {'non-printable', 'whitespace', 'other'})
        self.assertEqual(metrics['length'], 34)
        self.assertAlmostEqual(metrics['entropy'], 101.57032841892621)
        self.assertAlmostEqual(metrics['word_entropy'], 62.24279530886037)
        self.assertAlmostEqual(metrics['character_entropy'], 39.32753311006583)

        metrics = passwordmetrics.metrics('b4naNAf!ddler')
        self.assertEqual(metrics['words'], {'banana', 'fiddler'})
        self.assertEqual(metrics['used_groups'], {'uppercase', 'lowercase', 'punctuation', 'digits'})
        self.assertEqual(metrics['unused_groups'], {'non-printable', 'whitespace', 'other'})
        self.assertEqual(metrics['length'], 13)
        self.assertAlmostEqual(metrics['entropy'], 40.37060138599877)
        self.assertAlmostEqual(metrics['word_entropy'], 29.585966540441248)
        self.assertAlmostEqual(metrics['character_entropy'], 10.784634845557521)

        # Repeated words do not increase entropy
        # (Although they probably should add 1)
        self.assertEqual(passwordmetrics.metrics('battery')['entropy'],
                         passwordmetrics.metrics('batterybattery')['entropy'])

        # Repeated substitutions do not increase entropy:
        self.assertEqual(passwordmetrics.metrics('b4tterybattery')['entropy'],
                         passwordmetrics.metrics('b4tteryb4ttery')['entropy'])

        # Differenty substitutions *do' not increase entropy:
        self.assertNotEqual(passwordmetrics.metrics('b4tteryb4ttery')['entropy'],
                            passwordmetrics.metrics('b4ttery84ttery')['entropy'])


class TestCustomConfig(unittest.TestCase):

    def test_nonascii_chars(self):
        passwordmetrics.configure()
        # In 8-bit strings, ö is a part of Latin-1
        entropy, unkown = passwordmetrics._character_entropy(b'abcdefgh\xf6')
        self.assertEqual(unkown, set())

        # But this is not!
        entropy, unkown = passwordmetrics._character_entropy('abcdefgh\N{LATIN CAPITAL LETTER H WITH STROKE}')
        self.assertEqual(unkown, set('\N{LATIN CAPITAL LETTER H WITH STROKE}'))

        # If you want these to work, you have to make a custom configuration.
        if 'unicode' not in locals():
            unicode = str
        groups = {'lowercase': set(unicode(string.ascii_lowercase)),
                  'uppercase': set(unicode(string.ascii_uppercase)),
                  'digits': set(unicode(string.digits)),
                  'punctuation': set(unicode(string.punctuation)),
                  'whitespace': set(unicode(string.whitespace)),
                  'non-printable': set(unicode((i)) for i in range(128) if chr(i) not in string.printable),
                  'extras': set('åäöö\N{LATIN CAPITAL LETTER H WITH STROKE}'),
                  }

        passwordmetrics.configure(groups=groups)
        # This will still raise an error,
        entropy, unkown = passwordmetrics._character_entropy('abcdefghö\N{LATIN CAPITAL LETTER H WITH STROKE}')
        self.assertEqual(unkown, set())

    def test_custom_wordlist(self):
        # You might want to include non-english words in the word list.
        words = {}
        # Open a toy wordlist
        with open('docs/ordlista_sv.txt', 'rt', encoding='latin-1') as wordlist:
            for line in wordlist.readlines():
                word, entropy = line.strip().split(' ')
                words[word] = float(entropy)

        passwordmetrics.configure(words=words)

        # It won't find English words
        res = passwordmetrics.metrics('correcthorsebatterystapler')
        self.assertEqual(res['words'], set())

        # It will find some Swedish words
        res = passwordmetrics.metrics('korrekthästbatterihäftapparat')
        self.assertEqual(res['words'], set([u'batteri', u'korrekt', u'h\xe4ftapparat', u'h\xe4st']))


if __name__ == '__main__':
    unittest.main()


tests/test_password_entropy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_passwordmetrics
----------------------------------

Tests for `passwordmetrics` module.
"""
from __future__ import unicode_literals
import unittest
import passwordmetrics
import string


class TestPasswordMetrics(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        passwordmetrics.configure()
        cls.words = {'correct': 1, 'horse': 2, 'battery': 3, 'staple': 4, 'a': 1, 'incorrect': 1, 'bat': 1}
        
    def test_wordcount(self):
        words, rest = passwordmetrics._find_words('correcthorseb!wdbatterystaplerWd3t', self.words)
        self.assertEqual(words, set(['battery', 'horse', 'correct', 'staple']))
        self.assertEqual(rest, 'b!wdrWd3t')

        words, rest = passwordmetrics._find_words('abcdefgh', self.words)
        self.assertEqual(words, {'a'})
        self.assertEqual(rest, 'bcdefgh')        
                
    def test_character_entropy(self):
        entropy, groups, unkown = passwordmetrics._character_entropy('abcdefgh')
        self.assertAlmostEqual(entropy, 37.6035177451287)
        self.assertEqual(groups, {'lowercase'})

        entropy, groups, unkown = passwordmetrics._character_entropy('12345678')
        self.assertAlmostEqual(entropy, 26.5754247590989)
        self.assertEqual(groups, {'digits'})

        # Repeated characters count as one:
        entropy, groups, unkown = passwordmetrics._character_entropy('aAaAaAaA')
        self.assertAlmostEqual(entropy, 11.4008794362821)
        self.assertEqual(groups, {'lowercase', 'uppercase'})

        # As good as it gets for an 8 char password
        entropy, groups, unkown = passwordmetrics._character_entropy('xyFg98%!')
        self.assertAlmostEqual(entropy, 52.4367108134211)
        self.assertEqual(groups, {'lowercase', 'uppercase', 'digits', 'punctuation'})
        
    def test_substitutions(self):
        # This long password gets a reasonable score on characters alone
        self.assertAlmostEqual(passwordmetrics._character_entropy('Tr0ub4dor&3')[0], 65.54588851677637)
        # But much worse when considering it uses a word, even though it's misspelled and only 
        # appears once in the whole corpus (because I put it there).
        self.assertAlmostEqual(passwordmetrics.metrics('Tr0ub4dor&3')['entropy'], 31.332505617941614)
                
    def test_password_entropy(self):
        self.assertAlmostEqual(passwordmetrics.metrics('abcdefgh')['entropy'], 36.7646671442273)
        self.assertAlmostEqual(passwordmetrics.metrics('xyFg98%!')['entropy'], 48.9425059451315)
        self.assertAlmostEqual(passwordmetrics.metrics('correcthorsebatterystaple')['entropy'], 45.58216824444706)
        
    def test_password_metrics(self):
        metrics = passwordmetrics.metrics('correcthorseb!wdbatterystaplerWd3t')
        self.assertEqual(metrics['words'], {'et', 'stapler', 'i', 'correct', 'battery', 'horse'})
        self.assertEqual(metrics['used_groups'], {'uppercase', 'lowercase'})
        self.assertEqual(metrics['unused_groups'], {'punctuation', 'non-printable', 'whitespace', 'digits', 'other'})
        self.assertEqual(metrics['length'], 34)
        self.assertAlmostEqual(metrics['entropy'], 85.04455418142474)
        self.assertAlmostEqual(metrics['word_entropy'], 62.24279530886037)
        self.assertAlmostEqual(metrics['character_entropy'], 22.80175887256437)
        

class TestCustomConfig(unittest.TestCase):
        
    def test_nonascii_chars(self):
        passwordmetrics.configure()
        # In 8-bit strings, ö is a part of Latin-1
        entropy, groups, unkown = passwordmetrics._character_entropy(b'abcdefgh\xf6')
        self.assertEqual(unkown, set())
           
        # But this is not!
        entropy, groups, unkown = passwordmetrics._character_entropy('abcdefgh\N{LATIN CAPITAL LETTER H WITH STROKE}')
        self.assertEqual(unkown, set('\N{LATIN CAPITAL LETTER H WITH STROKE}'))
    
        # If you want these to work, you have to make a custom configuration.
        # I'll make one that uses all unicode characters even under Python 2.
        # This also get's rid of the warning.
        if 'unicode' not in locals():
            unicode = str
        groups = {'lowercase': set(unicode(string.ascii_lowercase)), 
                  'uppercase': set(unicode(string.ascii_uppercase)),
                  'digits': set(unicode(string.digits)),
                  'punctuation': set(unicode(string.punctuation)),
                  'whitespace': set(unicode(string.whitespace)),
                  'non-printable': set(unicode((i)) for i in range(128) if chr(i) not in string.printable), # non-printable
                  'extras': set('åäöö\N{LATIN CAPITAL LETTER H WITH STROKE}'),
                  }
        
        passwordmetrics.configure(groups=groups)
        # This will still raise an error, 
        entropy, groups, unkown = passwordmetrics._character_entropy('abcdefghö\N{LATIN CAPITAL LETTER H WITH STROKE}')
        self.assertEqual(unkown, set())
        # return to default config, just to be safe
        passwordmetrics.configure()        

    def test_verifyg(self):
        passwordmetrics.configure()
        groups = passwordmetrics.config['groups']
        for name, g in groups.items():
            for c in g:
                for n, x in groups.items():
                    if n == name:
                        continue
                    if c in x:
                        print("Group %s and group %s both have character %s" % (name, n, c))

if __name__ == '__main__':
    unittest.main()

utils/wordlist_compile.py
from io import open
from math import log
from collections import defaultdict

lemmas = defaultdict(lambda: 0)
chr_entropy = log(26, 2)

with open('ANC-written-count.txt', 'rt', encoding='latin-1') as infile:
    for line in infile.readlines():
        try:
            word, lemma, pos, count = line.strip().split('\t')
        except ValueError:
            # End of file
            break
        lemmas[lemma] += int(count)
    
if 'troubador' not in lemmas:
    # XKCD hack:
    lemmas['troubador'] = 1
    
most_common = max(lemmas.values())

for lemma, frequency in sorted(lemmas.items(), key=lambda x: x[1], reverse=True):
    # Calculate the words häufigkeit and use that as entropy.
    haufigkeit = 0.5 - log(frequency/most_common, 2)
    # If the entropy is *higher* than what you get with just characters, then it's
    # not a real word, and we skip it. Examples of this are the 'words' "abcd" and "fgh",
    # both in this database.
    if haufigkeit > len(lemma) * chr_entropy:
        continue
    print('%s %s' % (lemma, haufigkeit))
    

passwordmetrics/__init__.py
# -*- coding: utf-8 -*-
import math
import string
import collections
from io import open

__version__ = '1.1.dev0'

config = {}

def configure(groups=None, words=None, substitutions=None):
    global config

    # Define up the different character groups.
    if groups is None:
        # The default splits Latin-1 into seven different groups. The three last should be avoided, really.
        groups = {'lowercase': set(string.ascii_lowercase),
                  'uppercase': set(string.ascii_uppercase),
                  'digits': set(string.digits),
                  'punctuation': set(string.punctuation),
                  'whitespace': set(string.whitespace),
                  'non-printable': set(chr(i) for i in range(128) if chr(i) not in string.printable),
                  'other': set(chr(i) for i in range(128, 256)), # latin-1
                  }
    config['groups'] = groups

    # Configure the word list
    if words is None:
        # XXX Open as resource instead
        words = {}
        with open('passwordmetrics/wordlist_en.txt', 'rt', encoding='latin-1') as wordlist:
            for line in wordlist.readlines():
                word, entropy = line.strip().split(' ')
                words[word] = float(entropy)
    config['words'] = words

    # Set up common substitutions:
    if substitutions is None:
        substitutions = {'0': 'o', '1': 'i', '2': 'z', '3': 'e', '4': 'a',
                         '5': 's', '6': 'b', '7': 't', '8': 'b', '9': 'g',
                         '!': 'i', '#': '3', '$': 's', '&': 'g', '@': 'a',
                         '[': 'c', '(': 'c', '+': 't'}
    config['substitutions'] = substitutions


def _find_groups(pw):
    groups = config['groups']
    unused_groups = set(groups.keys())
    used_groups = set()

    for c in pw:
        for group in unused_groups:
            if c in groups[group]:
                used_groups.add(group)
                unused_groups.remove(group)
                break

    return used_groups, unused_groups


def _find_words(pw, words):
    """Returns the found words and the non-word characters"""
    max_len = max(map(len, words)) #longest word in the set of words
    max_chunks = max_len - 1
    words_found = set()
    chunks = []

    # Replace leet spellings
    substituted = ''.join([config['substitutions'].get(c, c).lower() for c in pw])

    # Find all possible words in the password
    for c in substituted:
        chunks = [chunk + c for chunk in chunks]
        chunks.insert(0, c)
        if len(chunks) > max_len:
            chunks.pop()
        for chunk in chunks:
            if chunk in words:
                words_found.add(chunk) #add to set of words

    # In "canotier" the above will find 'canotier', 'can', 'not', 'tier',
    # 'an', 'no' and 'a'. But only 'canotier' should count.
    rest = pw
    found = set()
    # We sort the words on length, and then alphabetically for consistency.
    # What order words are handled in can make a difference on entropy.
    # Should you for example have "2can0the" as a password it contains "can",
    # "not" and "the". If it picks "not" as the first word, this is the only
    # one that is found, but otherwise it will find "can" and "not". Possibly
    # this algorithm could be changed to try to find the most non-overlapping
    # words (and hence the lowest entropy) but this is good enough I think.
    remaining = substituted
    for word in sorted(words_found, key=lambda x: (len(x), x), reverse=True):
        word_length = len(word)
        while True:
            pos = remaining.find(word)
            if pos == -1:
                break
            rest = rest[:pos] + rest[pos+word_length:]
            remaining = remaining[:pos] + remaining[pos+word_length:]
            found.add(word)

    # Now check if any of these found words used character replacements.
    # In that case add those replacements to the "rest" to increase entropy.
    for word in found:
        start = 0
        while True:
            pos = substituted.find(word, start)
            if pos == -1:
                break
            original = pw[pos:pos+len(word)]
            for c in original:
                if c in config['substitutions']:
                    if c not in rest:
                        rest += c
            # Find all copies of the word
            start = pos + 1

    return found, rest


def _character_entropy(pw):
    if not pw:
        return 0, set()

    chars = set()
    unknown = set()
    for char in pw:
        if isinstance(char, int):
            char = chr(char)
        for name, group in config['groups'].items():
            if char in group:
                chars.update(group)
                break
        else:
            # The character is in none of the groups
            unknown.add(char)

    bit_per_word = math.log(len(chars), 2) # ie 128 characters would make 7 bits
    return len(set(pw)) * bit_per_word, unknown


def metrics(pw):
    all_words = config['words']
    used_groups, unused_groups = _find_groups(pw)
    found_words, rest = _find_words(pw, all_words)
    w = sum(all_words[word] for word in found_words)
    c, unknown = _character_entropy(rest)
    return {'entropy': w + c,
            'word_entropy': w,
            'character_entropy': c,
            'words': found_words,
            'used_groups': used_groups,
            'unused_groups': unused_groups,
            'length': len(pw),
            'unknown_chars': unknown,
            }
